src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/TdsParserStateObject.cs

```diff
@@ -1864,6 +1894,592 @@ internal bool TryReadPlpBytes(ref byte[] buff, int offset, int len, out int tota
+                    }
+                    else
+#endif
+                    {
+                        _networkPacketTaskSource.TrySetResult(null);
+                    }
+                    // Disable timeout timer on error
+                    SetTimeoutStateStopped();
+                    ChangeNetworkPacketTimeout(Timeout.Infinite, Timeout.Infinite);
+                }
+                else if (msecsRemaining == 0)
+                {
+                    // Got IO Pending, but we have no time left to wait
+                    // disable the timer and set the error state by calling OnTimeoutSync
+                    ChangeNetworkPacketTimeout(Timeout.Infinite, Timeout.Infinite);
+                    OnTimeoutSync();
+                }
+                // DO NOT HANDLE PENDING READ HERE - which is TdsEnums.SNI_SUCCESS_IO_PENDING state.
+                // That is handled by user who initiated async read, or by ReadNetworkPacket which is sync over async.
+            }
+            finally
+            {
+                if (!TdsParserStateObjectFactory.UseManagedSNI)
+                {
+                    if (!IsPacketEmpty(readPacket))
+                    {
+                        // Be sure to release packet, otherwise it will be leaked by native.
+                        ReleasePacket(readPacket);
+                    }
+                }
+                AssertValidState();
+            }
+        }
+
+        /// <summary>
+        /// Checks to see if the underlying connection is still alive (used by connection pool resiliency)
+        /// NOTE: This is not safe to do on a connection that is currently in use
+        /// NOTE: This will mark the connection as broken if it is found to be dead
+        /// </summary>
+        /// <param name="throwOnException">If true then an exception will be thrown if the connection is found to be dead, otherwise no exception will be thrown</param>
+        /// <returns>True if the connection is still alive, otherwise false</returns>
+        internal bool IsConnectionAlive(bool throwOnException)
+        {
+            Debug.Assert(_parser.Connection == null || _parser.Connection.Pool != null, "Shouldn't be calling IsConnectionAlive on non-pooled connections");
+            bool isAlive = true;
+
+            if (DateTime.UtcNow.Ticks - _lastSuccessfulIOTimer._value > CheckConnectionWindow)
+            {
+                if ((_parser == null) || ((_parser.State == TdsParserState.Broken) || (_parser.State == TdsParserState.Closed)))
+                {
+                    isAlive = false;
+                    if (throwOnException)
+                    {
+                        throw SQL.ConnectionDoomed();
+                    }
+                }
+                else if ((_pendingCallbacks > 1) || ((_parser.Connection != null) && (!_parser.Connection.IsInPool)))
+                {
+                    // This connection is currently in use, assume that the connection is 'alive'
```

Comment:
NIT, it's good to have!
```suggestion
                    SqlClientEventSource.Log.TryTraceEvent("TdsParserStateObject.IsConnectionAlive | Info | State Object Id {0}, This connection is currently in use, assume that the connection is 'alive'", _objectID);
```